pkg/controller/common.go

```diff
@@ -259,138 +268,56 @@ func (p *interPodAffinityFilter) filter(ctx context.Context, nodes []v1.Node) ([
-						continue
-					}
-					if value, exists := nodeInstance.GetLabels()[podAffinity.TopologyKey]; exists {
-						labelSet, exists := qualifiedLabelSet[podAffinity.TopologyKey]
-						if !exists {
-							labelSet = set{}
-						}
-						labelSet.add(value)
-						qualifiedLabelSet[podAffinity.TopologyKey] = labelSet
-					} else {
-						w.Logger().V(5).Infof("node (%s) doesn't have label value for topologyKey (%s)", nodeInstance.Name, podAffinity.TopologyKey)
-					}
+		for _, affinityTerm := range pod.Spec.Affinity.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
+			podNodes, replicaNodes := p.state.getQualifiedNodesForPodAffinityTerm(ctx, nodes, pod.Namespace, affinityTerm)
+			if isFirst {
+				qualifyingNodes = set{}
+				for podNode := range podNodes {
+					qualifyingNodes.add(podNode)
 				}
-
-				selector := labels.NewSelector()
-				for key, values := range qualifiedLabelSet {
-					labelValues := values.toStringSlice()
-					requirements, err := azureutils.CreateLabelRequirements(key, selection.In, labelValues...)
-					if err != nil {
-						w.Logger().Errorf(err, "failed to create label for key (%s) and values (%+v) :%v", key, labelValues, err)
-						continue
-					}
-					selector = selector.Add(*requirements)
+				for replicaNode := range replicaNodes {
+					qualifyingNodes.add(replicaNode)
 				}
-
-				// remove any candidate node which does not satisfy the pod affinity
-				for candidateNode := range candidateNodes {
-					if i, exists := nodeMap[candidateNode.(string)]; exists {
-						node := nodes[i]
-						nodeLabels := labels.Set(node.Labels)
-						if !selector.Matches(nodeLabels) {
-							w.Logger().V(5).Infof("Removing node (%s) from candidate nodes: node does not satisfy inter-pod-affinity (%v), label (%v)", candidateNode, podAffinity, selector)
-							candidateNodes.remove(candidateNode)
-						}
+				isFirst = false
+			} else {
+				for qualifyingNode := range qualifyingNodes {
+					if !podNodes.has(qualifyingNode) && !replicaNodes.has(qualifyingNode) {
+						qualifyingNodes.remove(qualifyingNode)
 					}
 				}
 			}
 		}
 	}
 
-	filteredNodes := []v1.Node{}
-	for replicaNode := range replicaNodes {
-		w.Logger().V(5).Infof("Adding node (%s) to candidate node list: node has a replica mount of qualifying pods", replicaNode)
-		candidateNodes.add(replicaNode)
-	}
-	for candidateNode := range candidateNodes {
-		if i, exists := nodeMap[candidateNode.(string)]; exists {
+	var filteredNodes []v1.Node
+	for qualifyingNode := range qualifyingNodes {
+		if i, exists := nodeMap[qualifyingNode.(string)]; exists {
 			filteredNodes = append(filteredNodes, nodes[i])
 		}
 	}
+	// Logging purpose
+	evictedNodes := make([]string, len(nodes)-len(filteredNodes))
+	i := 0
+	for _, node := range nodes {
+		if !qualifyingNodes.has(node.Name) {
+			evictedNodes[i] = node.Name
+			i++
+		}
+	}
+	w.Logger().V(5).Infof("nodes (%+v) filtered out by %s", evictedNodes, p.name())
```

Comment:
I wonder if we should use `V(10)` here since this would emit quick a lot of data. Using a quick search of the vendor sources, it seems this verbosity level is for emitting this level of detail.